crates/matrix-sdk/src/encryption/backups/mod.rs

```diff
@@ -0,0 +1,1157 @@
+        // so we can later on fall back to the correct `BackupState`.
+        let future = async {
+            self.set_state(BackupState::Enabling);
+
+            let olm_machine = self.client.olm_machine().await;
+            let olm_machine = olm_machine.as_ref().ok_or(Error::NoOlmMachine)?;
+
+            let decryption_key =
+                BackupDecryptionKey::from_base64(maybe_recovery_key).map_err(|e| {
+                    <serde_json::Error as serde::de::Error>::custom(format!(
+                        "Couldn't deserialize the backup recovery key: {e:?}"
+                    ))
+                })?;
+
+            // Let's try to see if there's a backup on the homeserver.
+            let current_version = self.get_current_version().await?;
+
+            let Some(current_version) = current_version else {
+                warn!("Tried to enable backups, but no backup version was found on the server.");
+                return Ok(false);
+            };
+
+            Span::current().record("backup_version", &current_version.version);
+
+            let backup_info: RoomKeyBackupInfo = current_version.algorithm.deserialize_as()?;
+            let stored_keys = olm_machine.backup_machine().get_backup_keys().await?;
+
+            if stored_keys.backup_version.as_ref() == Some(&current_version.version)
+                && self.are_enabled().await
+            {
+                // If we already have a backup enabled which is using the currently active
+                // backup version, do nothing but tell the caller using the return value that
+                // backups are enabled.
+                Ok(true)
+            } else if decryption_key.backup_key_matches(&backup_info) {
+                info!(
+                    "We have found the correct backup recovery key, storing the backup recovery \
+                     key and enabling backups"
+                );
+
+                // We're enabling a new backup, reset the `backed_up` flags on the room keys and
+                // remove any key/version we might have.
+                olm_machine.backup_machine().disable_backup().await?;
+
+                let backup_key = decryption_key.megolm_v1_public_key();
+                backup_key.set_version(current_version.version.to_owned());
+
+                // Persist the new keys and enable the backup.
+                olm_machine
+                    .backup_machine()
+                    .save_decryption_key(
+                        Some(decryption_key.to_owned()),
+                        Some(current_version.version.to_owned()),
+                    )
+                    .await?;
+                olm_machine.backup_machine().enable_backup_v1(backup_key).await?;
+
+                // If the user has set up the client to download any room keys, do so now. This
+                // is not really useful in a real scenario since the API to
+                // download room keys is not paginated.
```

Comment:
```suggestion
                // If the user has set up the client to download any room keys, do so now. This is
                // not really useful in a real scenario since the API to download room keys is not
                // paginated.
```